### PR TITLE
Adding OAuth MiddleWare

### PR DESCRIPTION
**Pull request by alexjoybc**
# Description

The OAuth MiddleWare can intercept outgoing http request and attach a Bearer Access Token int the header.
It helps decoupling the authentication mechanism from the usage of the dynamics API. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

I have tested locally with connection to dynamics CRM.

**Test Configuration**:

in `SearchRequestJob` replace with an existing search request ID:

```C#
var request = await _dynamicsApiClient.Get<SSG_SearchRequests>(
                $"ssg_searchrequests({Guid.NewGuid()})");
```

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/DynamicsAdapter/DynamicsAdapter.Web/SearchRequest/SearchRequestJob.cs [hunk 1]
```diff
@@ -32,15 +35,17 @@ public async Task Execute(IJobExecutionContext context)
...
-            _logger.LogDebug("Attempting to get token");
-            var token = await _oAuthApiClient.GetRefreshToken(CancellationToken.None);
-            _logger.LogInformation($"Successfully got token from provider, token expires in {token.ExpiresIn}");
+            _logger.LogDebug("Attempting to get search request from dynamics");
+            var request = await _dynamicsApiClient.Get<SSG_SearchRequests>(
+                $"ssg_searchrequests({Guid.NewGuid()})");
+            _logger.LogInformation($"Successfully got entity from dynamics");
 
             _logger.LogDebug("Attempting to post person search");
             var result = await _searchApiClient.SearchAsync(new PersonSearchRequest()
```
Comment (peggy-gov-bc-ca) [inline app/DynamicsAdapter/DynamicsAdapter.Web/SearchRequest/SearchRequestJob.cs]: add if (request !=null) to avoid NullException

Comment (alexjoybc) [reply app/DynamicsAdapter/DynamicsAdapter.Web/SearchRequest/SearchRequestJob.cs]: I'm not using request anywhere now, so don;t need to check for null ;-)

File: app/DynamicsAdapter/DynamicsAdapter.Web/Auth/TokenService.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            var tokenString = await _distributedCache.GetStringAsync(this.token_key);
+            if (String.IsNullOrEmpty(tokenString)) return null;
+            return JsonConvert.DeserializeObject<Token>(tokenString);
+        }
+
+        private async Task<Token> RefreshTokenAsync(CancellationToken cancellationToken)
+        {
+            var token = await _oAuthApiClient.GetRefreshToken(cancellationToken);
+            await _distributedCache.SetStringAsync(this.token_key, JsonConvert.SerializeObject(token), new DistributedCacheEntryOptions() { AbsoluteExpirationRelativeToNow = new TimeSpan(0, 0, token.ExpiresIn - 10) }, cancellationToken);
```
[the remark this answers is not in the document]
Comment (alexjoybc) [reply app/DynamicsAdapter/DynamicsAdapter.Web/Auth/TokenService.cs]: Agree

File: app/DynamicsAdapter/DynamicsAdapter.Web/SearchRequest/SearchRequestJob.cs (hunk 1 above)
Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/DynamicsAdapter.Web/SearchRequest/SearchRequestJob.cs]: request.SSG_PersonGivenName,

Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/DynamicsAdapter.Web/SearchRequest/SearchRequestJob.cs]: FirstName = request.SSG_PersonGivenName,
                LastName = request.SSG_PersonSurname,
                DateOfBirth = request.SSG_PersonBirthDate
those are not used anymore?

**Pull request merged**